src/Demo/ColorPicker/Form/Type/Step/Step2Type.php

```diff
@@ -0,0 +1,57 @@
+<?php
+
+namespace App\Demo\ColorPicker\Form\Type\Step;
+
+use App\Demo\ColorPicker\Form\Data\ColorPickerDto;
+use Symfony\Component\Form\AbstractType;
+use Symfony\Component\Form\Extension\Core\Type\ChoiceType;
+use Symfony\Component\Form\FormBuilderInterface;
+use Symfony\Component\OptionsResolver\OptionsResolver;
+
+class Step2Type extends AbstractType
+{
+    public function buildForm(FormBuilderInterface $builder, array $options): void
+    {
+        /** @var ColorPickerDto $colorPicker */
+        $colorPicker = $builder->getData();
+
+        assert(is_string($colorPicker->mainColor), 'Main color must be a string');
+
+        $gradients = [
+            'Lighter' => $this->adjustBrightness($colorPicker->mainColor, 40),
+            'Original' => $colorPicker->mainColor,
+            'Darker' => $this->adjustBrightness($colorPicker->mainColor, -40),
+        ];
+
+        $builder->add('gradientColor', ChoiceType::class, [
+            'label' => 'Gradient Color',
+            'choices' => $gradients,
+            'choice_label' => function ($choice, $value) {
+                return $value . ' (' . $choice . ')';
+            },
+        ]);
+    }
+
+    public function configureOptions(OptionsResolver $resolver): void
+    {
+        $resolver->setDefaults([
+            'data_class' => ColorPickerDto::class,
+            'inherit_data' => true,
```

Comment:
The reason the PRE_SET_DATA and POST_SET_DATA events are not dispatched is due to this specific option, and this behavior is expected outside of a FormFlow context. If you need access to the parent data within a step, you must explicitly pass it through the addStep() options.

Or use the alternative you used first (via custom options).